### DISCUSSION TITLE
Built-In Service Discovery / Ingress

### DISCUSSION THREAD

**Discussion by Danielku15** [dotnet/yarp#1872] [Ideas] [answered]
Hi everyone. 

It would be great if yarp would provide some built-in Service Discovery mechanisms without the need of relying on K8S, Consul, Eureka etc. 

Maybe Service **Discovery** is not really the right term. What I'm rather thinking is a simplified version of the [Tunneling feature discussed here](https://github.com/microsoft/reverse-proxy/issues/16180). The usecase is like:

I am providing my applications a configuration containing: 
* Where to find YARP
* How to authenticate against YARP
* To which service type + cluster (composite cluster id) they belong. 

Whenever an application is started in a dynamic scaling environment (no matter whether K8S, custom orchestration like manual installation, Ansible, Azure App Services) it will connect to YARP and register itself in the cluster and perform a regular health-check & keep-alive. 

At this point everything is as usual: YARP will look in the list of destinations which were dynamically registered through the process above. 

Has this been discussed within the YARP community already? I was thinking that with a custom API (or maybe websocket channel) I guess I could handle this also as a custom YARP extension but something more common/standardized/built-in is always preferrable 😁

**Comment by Tratcher** [comment, marked as answer]
Self-registering destinations hasn't come up. This could be done as an implementation of IProxyConfigProvider that was updated from a listening endpoint. The protocol doesn't matter (HTTP/REST, WS, etc.). You're right to call out that this endpoint would have to be secured.

The biggest downside is that it also requires a client on each node being proxied to, written in the language of that client (Java, node, .Net, etc). For that reason you'd want to keep the interaction as simple as possible, likely some REST HTTP calls. That's one reason it's generally preferable to rely on an orchestrator for the discovery. Another is that you'd need a garbage collection mechanic for destinations that go offline for too long and can't self-unregister.

**Reply by Tratcher** [reply]
Feel free to do some prototyping and show us what you come up with.

**Reply by Danielku15** [reply]
Thanks for the info that there was no discussion at least until now. In my ongoing prototyping the handshake is very simple as we're not expecting thousands of destinations: 

1. Yarp has a predefined list of routes and clusters for all known services (we only need dynamic destinations). 
2. The destinations connect through websocket to YARP and pass along a registration request consisting of the cluster to which they want ot be added (has to be known by the destination by configuration) and the base URL where the destination is exposed (host part of the URL is replaced with the remote ip address)
3. This connection is kept alive by the ASP.net core infrastructure through the `KeepAliveInterval`. 
4. In the request pipeline of the websocket connection we add the destination to the cluster and then keep checking that the the WebSocket connection did not drop.
5. Once the request pipeline completes the destination is unregistered. 
6. It is the destinations duty to reconnect and keep the link alive. 

Most of the code is fully independent from YARP until now and not more than 200 lines of code. The only actual missing part I need to dive into the docs and extensibility points is how to exactly add/remove destinations dynamically to/from pre-configured clusters so that they are considered by YARP. Thanks for the hint with `IProxyConfig` but I also found `IClusterDestinationsUpdater` which sounds very promising for my use case. Seems I could just call `UpdateAllDestinations` with an updated state.

**Reply by Tratcher** [reply]
Adding and removing destinations should be managed through IProxyConfigProvider. IClusterDestinationsUpdater is more about filtering out unhealthy destinations based on health checks.

**Reply by Danielku15** [reply]
Thanks for the input. A few hours later I have an operational and running version of my concept now. Roughly 150 Lines of Code for the client part, Roughtly 600 Lines of Code for the server part. It is at this point not super resilient but the mechanism of maintaining the available destinations of a cluster through WebSocket channels works fine. 

The only pain point was that many things of the built-in configuration loading are internal. I still want to load the main part of my route and cluster configs from the `IConfiguration` but cannot utilize the built-in `ConfigurationConfigProvider` for that. Maybe I file an Issue to make this provider part of the public API so others can inherit, extend, and use it for their purposes.

**Reply by Tratcher** [reply]
Awesome. Can you show how you did it, perhaps as a github repo?

**Reply by Danielku15** [reply]
Yes, still need to find some time to extract the relevant parts into a standalone repository. I made the extensions directly as part of a product we're developing. 😉

**Reply by eyalg-msft** [reply]
Hi @Danielku15, 
We also use YARP and we need a dynamic update (periodic) of our destination list.

This is what we did:

1.Added a BackgroundService which I called "ConfigurationUpdaterBackgroundService" that runs every 3 minutes.

2.On the ConfigureServices when creating our new BackgroundService we use the service provider to get the IProxyConfigProvider and pass it to the background service constructor.

`services.AddHostedService<ConfigurationUpdaterBackgroundService>(serviceProvider => new ConfigurationUpdaterBackgroundService(proxyBuilder, ServiceInstanceFinder, Configuration, serviceProvider.GetRequiredService<IProxyConfigProvider>()));`

3.The BackgroundService will now periodically check for updates in the list (we have another service to get it), and if needed it wil call the Update function of YARP:

`this.reverseProxyInMemoryConfigProvider.Update(routes, clusters);`

I would be happy to hear your thoughts of this solution. Is this the right way to do it? do you have a recommendation to do it diffrently?

**Reply by Danielku15** [reply]
I quickly put a standalone project together with our PoC. As announced: not super sophisticated at this point but it does its job 😁 It also shows a bit the pain point of not being able to use the default config classes and loaders. 

https://github.com/Danielku15/yarp-service-discovery
